### ISSUE TITLE
graceful login error handling?

### ISSUE THREAD

**Issue by vladimir-mencl-eresearch** [chauth/confluence_http_authenticator#33]
Hi,

I have run into an issue on our tiny (50 user) Confluence server.

We were out of user licenses.  And a new user trying to log in via the Shibboleth Authenticator was getting weird behavior - just being redirected back to the Discovery Service (starting a new Shibboleth login).

I was first suspecting an issue with cookies being blocked - and it took me a while before I found this in the logs:

````
2015-11-18 07:13:48,681 ERROR [http-8080-7] [confluence.authentication.shibboleth.RemoteUserAuthenticator] doInTransaction Failed to create user 'colin001@virtualhome.tuakiri.ac.nz'!
bucket.user.LicensingException: You are not licensed to add any more users to this installation of Confluence. Please contact sales@atlassian.com
        at com.atlassian.confluence.user.DefaultUserAccessor.userWillBeCreated(DefaultUserAccessor.java:213)
        at com.atlassian.confluence.user.DefaultUserAccessor.createUser(DefaultUserAccessor.java:198)
        at sun.reflect.GeneratedMethodAccessor1523.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:182)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:149)
        at com.atlassian.spring.interceptors.SpringProfilingInterceptor.invoke(SpringProfilingInterceptor.java:16)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:106)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
        at com.sun.proxy.$Proxy40.createUser(Unknown Source)
        at shibauth.confluence.authentication.shibboleth.RemoteUserAuthenticator$3.doInTransaction(RemoteUserAuthenticator.java:1243)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:128)
        at shibauth.confluence.authentication.shibboleth.RemoteUserAuthenticator.createUser(RemoteUserAuthenticator.java:1240)
        at shibauth.confluence.authentication.shibboleth.RemoteUserAuthenticator.createUser(RemoteUserAuthenticator.java:361)
        at shibauth.confluence.authentication.shibboleth.RemoteUserAuthenticator.getUser(RemoteUserAuthenticator.java:971)
        at com.atlassian.seraph.filter.SecurityFilter.doFilter(SecurityFilter.java:125)
````

I'm not sure what the right fix would be - but can we do some graceful handling so that errors like this do get reported to the user?  I.e., redirect to an error page instead of just trying a new login?

Cheers,
Vlad

PS: Full disclosure: this system is running an ancient version of Confluence (4.3.7) with an ancient version of the remoteAuthenticator (2.1.16) ... but I'm expecting this has not changed in the current version.

**Comment by garysweaver** [comment]
@vladimir-mencl-eresearch Hey, sure that sounds great!

If you have time for some coding, try letting that exception bubble up from ```userAccessor.createUser(username);``` by catching something (or things) other than Throwable or Exception and see if that works. It will be a few days before I'll be able to take a look at it otherwise. It looks like that should work, since it is a RuntimeException: https://docs.atlassian.com/confluence/latest/bucket/user/LicensingException.html

**Comment by garysweaver** [comment]
Released attempted fix for this in 2.7.0: https://github.com/chauth/confluence_http_authenticator/blob/master/releases/remoteUserAuth-2.7.0.jar?raw=true

Let me know if that helps.
